vital/tests/test_const_iterator.cxx

```diff
@@ -0,0 +1,409 @@
+TEST( const_iterator, prefix_increment )
+{
+  typedef vital::const_iterator<unsigned int> iter_t;
+
+  unsigned int v = -1;
+  iter_t::next_value_func_t nvf = [&]()->iter_t::reference{
+    ++v;
+    return v;
+  };
+
+  iter_t it( nvf );
+  ASSERT_EQ( *it, 0 );
+  ASSERT_EQ( v, 0 );
+
+  ASSERT_EQ( *++it, 1 );
+  ASSERT_EQ( v, 1 );
+  ASSERT_EQ( *++it, 2 );
+  ASSERT_EQ( v, 2 );
+  ASSERT_EQ( *++it, 3 );
+  ASSERT_EQ( v, 3 );
+  ASSERT_EQ( *++it, 4 );
+  ASSERT_EQ( v, 4 );
+}
+
+// Test postfix incrementing operator over integer sequence generator.
+TEST( const_iterator, postfix_increment )
+{
+  typedef vital::const_iterator<unsigned int> iter_t;
+
+  // The postfix operation requires that the generator function return unique
+  // references, so we make an array to iterate over.
+  unsigned int a[] = {0, 1, 2, 3};
+  iter_t::next_value_func_t nvf = [&]()->iter_t::reference{
+    static size_t i = 0;
+    return a[i++];
+  };
+
+  iter_t it( nvf );
+  ASSERT_EQ( *it, 0 );
+  ASSERT_EQ( *it++, 0 );
+  ASSERT_EQ( *it++, 1 );
+  ASSERT_EQ( *it++, 2 );
+  ASSERT_EQ( *it, 3 );
+}
+
+// Test iterating over pointers and using arrow operator.
+TEST( const_iterator, pointer_iteration_arrow_operator )
+{
+  // Simple wrapper structure to test arrow operations.
+  typedef struct int_container {
+    int i;
+    int_container( int v )
+      : i( v )
+    {}
+  } int_container;
+  typedef vital::const_iterator< int_container > iter_t;
+
+  int_container a[] = { int_container( 0 ),
+                        int_container( 1 ),
+                        int_container( 2 ) };
+  iter_t it( [&] () ->iter_t::reference {
+    static size_t i = 0;
+    return a[i++];
+  } );
+
+  ASSERT_EQ( it->i, 0 );
+  ASSERT_EQ( it++->i, 0 );
+  ASSERT_EQ( it->i, 1 );
+  //++it;
+  //ASSERT_EQ( it->i, 2 );
+  ASSERT_EQ( (++it)->i, 2 );
+}
+
+// Test that two iterators are equal at points where their current values are
+// equal or both are at the end of iteration because the generation function
+// raise a stop iteration exception.
+TEST( const_iterator, it_equality )
+{
+  using namespace std;
+
+  typedef vital::const_iterator<int> iter_t;
+  int a[] = { 10, 11, 12, 13 };
+
+  cout << "Creating iterators" << endl;
+  iter_t it1( [&]()->iter_t::reference{
+    static size_t i = 0;
+    if( i == 4 )
+    {
+      cout << "Raising stop iteration" << endl;
+      throw vital::stop_iteration_exception();
+    }
+    cout << "returning a[" << i << "]" << endl;
+    return a[i++];
+  } );
+  iter_t it2( [&]()->iter_t::reference{
+    static size_t i = 0;
+    if( i == 4 )
+    {
+      cout << "Raising stop iteration" << endl;
+      throw vital::stop_iteration_exception();
+    }
+    cout << "returning a[" << i << "]" << endl;
+    return a[i++];
+  } );
+
+  cout << "testing first values" << endl;
+  ASSERT_EQ( *it1, 10 );
+  ASSERT_EQ( *it2, 10 );
+
+  // Initial values, equal, iterators should be equal.
+  ASSERT_TRUE( it1 == it2 );
```

Comment:
Why are these not using `EXPECT_EQ` and `EXPECT_NE`?

Reply:
I thought this form would be more explicit in the equality operation that I was trying to test.

Reply:
I suppose that's true. OTOH, it will not print the operands if the assertion fails.

Reply:
However, with these asserts, the "operand" print is:
```
97:       Expected: it1                                                                                                                  
97:       Which is: 56-byte object <00-8B 41-00 00-00 00-00 F0-BE 7C-D4 FC-7F 00-00 00-00 00-00 00-00 00-00 CC-E3 40-00 00-00 00-00 AA-E3
 40-00 00-00 00-00 F8-BE 7C-D4 FC-7F 00-00 00-00 00-00 00-00 00-00>                                                                      
97: To be equal to: it2                                                                                                                  
97:       Which is: 56-byte object <00-8B 41-00 00-00 00-00 F0-BE 7C-D4 FC-7F 00-00 CC-E3 40-00 00-00 00-00 2A-E5 40-00 00-00 00-00 08-E5
 40-00 00-00 00-00 F4-BE 7C-D4 FC-7F 00-00 00-C0 7C-D4 FC-7F 00-00>                                                                      
```
So that's just about as helpful as just saying "yeah, they weren't equal".